kitsune/wiki/views.py

```diff
@@ -483,94 +483,123 @@ def edit_document(request, document_slug, revision_id=None):
+    """Edit document metadata."""
+    user = request.user
 
-                topics = []
-                for t in post_data.getlist("topics"):
-                    topics.append(int(t))
-                post_data.setlist("topics", topics)
+    doc = get_visible_document_or_404(
+        user,
+        locale=request.LANGUAGE_CODE,
+        slug=document_slug,
+        look_for_translation_via_parent=True,
+        return_parent_if_no_translation=True,
+    )
 
-                doc_form = DocumentForm(
-                    post_data,
-                    instance=doc,
-                    can_archive=can_archive,
-                    can_edit_needs_change=can_edit_needs_change,
-                )
-                if doc_form.is_valid():
-                    # Get the possibly new slug for the imminent redirection:
-                    try:
-                        doc = doc_form.save(None)
-                    except (TitleCollision, SlugCollision) as e:
-                        # TODO: .add_error() when we upgrade to Django 1.7
-                        errors = doc_form._errors.setdefault("title", ErrorList())
-                        message = "The {type} you selected is already in use."
-                        message = message.format(
-                            type="title" if isinstance(e, TitleCollision) else "slug"
-                        )
-                        errors.append(_(message))
-                    else:
-                        # Do we need to rebuild the KB?
-                        _maybe_schedule_rebuild(doc_form)
+    if not doc.allows(user, "edit") or not user.is_staff:
+        raise PermissionDenied
 
-                        return HttpResponseRedirect(
-                            urlparams(
-                                reverse("wiki.edit_document", args=[doc.slug]), opendescription=1
-                            )
-                        )
-                disclose_description = True
-            else:
-                raise PermissionDenied
-        elif which_form == "rev":
-            if doc.allows(user, "create_revision"):
-                rev_form = RevisionForm(request.POST)
-                rev_form.instance.document = doc  # for rev_form.clean()
-                if rev_form.is_valid():
-                    _save_rev_and_notify(rev_form, user, doc, base_rev=rev)
-                    if "notify-future-changes" in request.POST:
-                        EditDocumentEvent.notify(request.user, doc)
+    if doc.locale != request.LANGUAGE_CODE:
+        # We've fallen back to the parent, since no visible translation existed.
+        url = reverse("wiki.translate", locale=request.LANGUAGE_CODE, args=[document_slug])
+        return HttpResponseRedirect(url)
 
-                    return HttpResponseRedirect(
-                        reverse("wiki.document_revisions", args=[document_slug])
-                    )
+    can_edit_needs_change = doc.allows(user, "edit_needs_change")
```

Comment:
There's a big copy/paste block that's shared between the two views. Is it possible to make this a bit more DRY?

Reply:
Can do.